---
title: "RiboViz simASite Regressions"
author: "Amanda Mok"
date: "01/14/2020"
output: html_document
---

```{r global, echo=F}
knitr::opts_chunk$set(cache=T, echo=F, fig.width=12, warning=F)

library(foreach)
library(pscl)
library(ggplot2)
library(patchwork)
library(gridExtra)
library(kableExtra)
```

```{r load-data}
source("~/footprint-bias/scripts/prep_data.R")

# specify filenames
data_dir <- "~/footprint-bias/sample_data"
transcript_fa_fname <- file.path(data_dir, "riboViz_transcripts_18cds15.fa")
transcript_length_fname <- file.path(data_dir, "riboViz_transcripts_18cds15_lengths.txt")
offsets_fname <- file.path(data_dir, "Asite_rules.txt")
sam_fname <- file.path(data_dir, "riboViz_simASite_footprints.sam")

# initialize regression framework
print("initializing data.frame (2nt) ...")
system.time({
  regression_data_2nt <- init_data(transcript_fa_fname, transcript_length_fname, bias_length=2)
})
print("counting footprints (2nt) ...")
system.time({
  regression_data_2nt <- count_footprints(sam_fname, transcript_length_fname, 
                                          offsets_fname, regression_data_2nt)
})
print("initializing data.frame (3nt) ...")
system.time({
  regression_data_3nt <- init_data(transcript_fa_fname, transcript_length_fname, bias_length=3)
})
print("counting footprints (3nt) ...")
system.time({
  regression_data_3nt <- count_footprints(sam_fname, transcript_length_fname, 
                                          offsets_fname, regression_data_3nt)
})
save(regression_data_2nt, regression_data_3nt, file="../sample_data/riboViz_simASite_regression_data.Rda")
```

### regression models

* footprint count ~ exp(transcript + Asite + Psite + Esite + 5' digest length + 3' digest length + 5' bias sequence + 3' bias sequence)

* distributions for footprint count:

+ poisson: $var(x) = E[x]$

+ quasi-poisson: $var(x) = \phi \cdot E[x]$

+ negative binomial: $var(x) = E[x] + \frac{(E[x])^2}{\phi}$

* bias parameters: modeled as 2nt or 3nt sequences

```{r fit-all-data}
regression <- formula(count ~ transcript + A + P + E + d5 + d3 + f5 + f3)
all_fits_2nt <- list(poisson=glm(regression, data=regression_data_2nt, family="poisson"),
                     quasipois=glm(regression, data=regression_data_2nt, family="quasipoisson"),
                     negbin=MASS::glm.nb(regression, data=regression_data_2nt))
all_fits_3nt <- list(poisson=glm(regression, data=regression_data_3nt, family="poisson"),
                     quasipois=glm(regression, data=regression_data_3nt, family="quasipoisson"),
                     negbin=MASS::glm.nb(regression, data=regression_data_3nt))

# data frame with residuals
residuals_2nt <- data.frame(fit=factor(rep(names(all_fits_2nt),
                                           each=nrow(regression_data_2nt)),
                                       levels=names(all_fits_2nt)),
                            regression_data_2nt,
                            predicted=unlist(lapply(all_fits_2nt, function(x) predict(x, type="response"))), 
                            residual=unlist(lapply(all_fits_2nt, function(x) residuals(x, type="response"))))
residuals_summed_2nt <- data.frame(aggregate(count ~ transcript + cod_idx + fit, data=residuals_2nt, FUN=sum),
                                   predicted=aggregate(predicted ~ transcript + cod_idx + fit, 
                                                       data=residuals_2nt, FUN=sum)$predicted,
                                   residuals=aggregate(residual ~ transcript + cod_idx + fit,
                                                       data=residuals_2nt, FUN=sum)$residual)
residuals_3nt <- data.frame(fit=factor(rep(names(all_fits_3nt),
                                           each=nrow(regression_data_3nt)),
                                       levels=names(all_fits_3nt)),
                            regression_data_3nt,
                            predicted=unlist(lapply(all_fits_3nt, function(x) predict(x, type="response"))), 
                            residual=unlist(lapply(all_fits_3nt, function(x) residuals(x, type="response"))))
residuals_summed_3nt <- data.frame(aggregate(count ~ transcript + cod_idx + fit, data=residuals_3nt, FUN=sum),
                                   predicted=aggregate(predicted ~ transcript + cod_idx + fit, 
                                                       data=residuals_3nt, FUN=sum)$predicted,
                                   residuals=aggregate(residual ~ transcript + cod_idx + fit,
                                                       data=residuals_3nt, FUN=sum)$residual)
save(all_fits_2nt, all_fits_3nt, residuals_2nt, residuals_summed_2nt, residuals_3nt, residuals_summed_3nt,
     file="../sample_data/riboViz_simASite_regression.Rda")
```

### plot fits
```{r plot-model-predictions, fig.height=12}
# model predictions
pred_2nt <- ggplot(residuals_summed_2nt) + geom_line(aes(x=cod_idx, y=count)) + 
  geom_line(aes(x=cod_idx, y=predicted, color=fit)) + 
  facet_grid(fit~transcript, scales="free") + 
  theme_bw() + xlab("codon position") + ylab("predicted") + ggtitle("Regression predictions: 2nt bias seq")
pred_3nt <- ggplot(residuals_summed_3nt) + geom_line(aes(x=cod_idx, y=count)) + 
  geom_line(aes(x=cod_idx, y=predicted, color=fit)) + 
  facet_grid(fit~transcript, scales="free") + 
  theme_bw() + xlab("codon position") + ylab("predicted") + ggtitle("Regression predictions: 3nt bias seq")
pred_2nt / pred_3nt
```

* Model predictions better recapitulate original data if we model recovery biases with sequences that are 3 nucleotides in length (versus 2 nucleotides).

* Not much visual difference between GLM families (i.e. poisson vs. quasi-poisson vs. negative binomial).

```{r plot-model-residuals, fig.height=12}
# model residuals
resid_2nt <- ggplot(residuals_summed_2nt) + geom_line(aes(x=cod_idx, y=residuals, color=fit)) + 
  facet_grid(fit~transcript, scales="free") + 
  theme_bw() + xlab("codon position") + ylab("residual") + ggtitle("Regression residuals: 2nt bias seq")
resid_3nt <- ggplot(residuals_summed_3nt) + geom_line(aes(x=cod_idx, y=residuals, color=fit)) + 
  facet_grid(fit~transcript, scales="free") + 
  theme_bw() + xlab("codon position") + ylab("residual") + ggtitle("Regression residuals: 3nt bias seq")
resid_2nt / resid_3nt
```

* Residuals are much smaller in magnitude with 3nt bias sequence models.

* Not much visual difference between GLM families. Models tend to fail in similar places and similar directions (i.e. only vary in magnitude).

### evaluate model fits

```{r evaluate-model-fits}
model_eval <- data.frame(fit=c(paste(names(all_fits_2nt), "(2nt)"), paste(names(all_fits_3nt), "(3nt)")),
                         deviance=c(sapply(all_fits_2nt, deviance), sapply(all_fits_3nt, deviance)),
                         AIC=c(sapply(all_fits_2nt, AIC), sapply(all_fits_3nt, AIC)),
                         dispersion=c(sapply(all_fits_2nt, function(x) summary(x)$dispersion),
                                      sapply(all_fits_3nt, function(x) summary(x)$dispersion)),
                         total_error=c(sapply(all_fits_2nt, 
                                              function(x) sum(abs(residuals(x, type="response")))),
                                       sapply(all_fits_3nt,
                                              function(x) sum(abs(residuals(x, type="response"))))))
model_order <- as.character(model_eval$fit)
model_eval$fit <- factor(model_eval$fit, levels=model_order)
kable_styling(column_spec(knitr::kable(model_eval, row.names=F), 
                          column=1:ncol(model_eval), border_left=T, border_right=T))
```

* Note: quasi-Poisson is not a "real" distribution; therefore, does not have a likelihood/AIC

```{r plot-model-fits}
ggplot(model_eval, aes(fit, deviance)) + geom_bar(stat="identity") +
  theme_bw() + xlab("model") + ylab("") + ggtitle("deviance") + 
  theme(axis.text.x=element_text(angle=90, hjust=1))
ggplot(model_eval, aes(fit, total_error)) + geom_bar(stat="identity") +
  theme_bw() + xlab("model") + ylab("") + ggtitle("Total error") +
  theme(axis.text.x=element_text(angle=90, hjust=1))

print("testing... poisson vs. negative binomial regression (2nt)")
pscl::odTest(all_fits_2nt$negbin)
print("testing... poisson vs. negative binomial regression (3nt)")
pscl::odTest(all_fits_3nt$negbin)
```

* 3nt models much lower deviance and total error than 2nt models

* Negative binomial models much lower deviance than poisson and quasi-poisson models, but moderately higher total error (also demonstrated by residual plots above)

* Likelihood ratio test indicate negative binomial regressions fit data much better than poisson regressions

### violin plots of model residuals

```{r residuals}
# violin plots
resid_violin_2nt <- ggplot(residuals_2nt, aes(fit, residual)) + geom_violin(aes(fill=fit)) + 
  theme_bw() + ggtitle("Regression residuals: 2nt bias seq")
resid_violin_2nt_subset <- ggplot(residuals_2nt, aes(fit, residual)) + geom_violin(aes(fill=fit)) + 
  theme_bw() + coord_cartesian(ylim=c(-100, 500))
resid_violin_3nt <- ggplot(residuals_3nt, aes(fit, residual)) + geom_violin(aes(fill=fit)) + 
  theme_bw() + ggtitle("Regression residuals: 3nt bias seq")
resid_violin_3nt_subset <- ggplot(residuals_3nt, aes(fit, residual)) + geom_violin(aes(fill=fit)) + 
  theme_bw() + coord_cartesian(ylim=c(-100, 500))
(resid_violin_2nt + resid_violin_3nt) / (resid_violin_2nt_subset + resid_violin_3nt_subset)
```

* Not much visual difference between poisson and quasi-poisson models

* Residuals from negative binomial models tend to have wider range, but more observations at 0

### mean-variance plots for residuals

```{r residuals-mean-var, fig.height=12}
residuals_2nt$residual_sq <- residuals_2nt$residual^2
residuals_3nt$residual_sq <- residuals_3nt$residual^2
nb_mean_var_2nt <- data.frame(x=seq(from=1, to=round(max(residuals_2nt$predicted))))
nb_mean_var_2nt$y <- nb_mean_var_2nt$x + nb_mean_var_2nt$x^2/summary(all_fits_2nt[["negbin"]])$dispersion
nb_mean_var_3nt <- data.frame(x=seq(from=1, to=round(max(residuals_3nt$predicted))))
nb_mean_var_3nt$y <- nb_mean_var_3nt$x + nb_mean_var_3nt$x^2/summary(all_fits_3nt[["negbin"]])$dispersion
fit_colors <- hcl(h=seq(15, 375, length=length(all_fits_2nt)+1), l=65, c=100)
mean_var_2nt <- ggplot(residuals_2nt, aes(x=predicted, y=residual_sq, color=fit)) + geom_jitter(alpha=0.2) + 
  theme_bw() + xlab("mean") + ylab("variance") + ggtitle("2nt") +
  geom_abline(slope=1, intercept=0, color=fit_colors[1]) +
  geom_abline(slope=summary(all_fits_2nt[["quasipois"]])$dispersion, intercept=0, color=fit_colors[2]) +
  geom_line(data=nb_mean_var_2nt, mapping=aes(x=x, y=y), color=fit_colors[3])
mean_var_2nt_subset <- mean_var_2nt + xlim(0, 1000) + ylim(0, 1e7)
mean_var_3nt <- ggplot(residuals_3nt, aes(x=predicted, y=residual_sq, color=fit)) + geom_jitter(alpha=0.2) + 
  theme_bw() + xlab("mean") + ylab("variance") + ggtitle("3nt") +
  geom_abline(slope=1, intercept=0, color=fit_colors[1]) +
  geom_abline(slope=summary(all_fits_3nt[["quasipois"]])$dispersion, intercept=0, color=fit_colors[2]) +
  geom_line(data=nb_mean_var_3nt, mapping=aes(x=x, y=y), color=fit_colors[3])
mean_var_3nt_subset <- mean_var_3nt + xlim(0, 1000) + ylim(0, 1e6)
(mean_var_2nt + mean_var_2nt_subset) / (mean_var_3nt + mean_var_3nt_subset)
```

* Each point represents a (transcript, codon, 5' digest length, 3' digest length)

* Mean is model prediction, variance is square of residual

* Residuals from 2nt models very over-dispersed, not quite captured by quasi-poisson or negative binomial models

* Dispersion better captured by negative binomial in 3nt models

### residual distributions, by predictors

```{r residuals-vs-predictors, fig.height=8}
plot_residuals_predictors <- function(resid_table_2nt, resid_table_3nt, predictor, plotTitle) {
  # 3 plots:
  #     barplot of # observations per category of predictor
  #     boxplot of residuals by category of predictor, colored by model
  #     boxplot of actual footprint count per category of predictor
  ## resid_table_2nt: data.frame with column names "fit", "residual", "count", and predictor (2nt models)
  ## resid_table_3nt: data.frame with column names "fit", "residual", "count", and predictor (3nt models)
  ## predictor: character; corresponds to column name in resid_table
  ##            predictor must be a factor
  ## title: character; string for ggplot2::ggtitle()
  require(ggplot2)
  require(patchwork)
  # 2nt plots
  pred_barplot_2nt <- ggplot(resid_table_2nt, aes_string(predictor)) + geom_bar() + 
    theme_bw() + xlab("") + ylab("# obs") + ggtitle(plotTitle) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1))
  pred_residplot_2nt <- ggplot(resid_table_2nt, aes_string(predictor, "residual")) + 
    geom_boxplot(aes(color=fit)) + theme_bw() + xlab("") + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1))
  pred_counts_2nt <- ggplot(resid_table_2nt, aes_string(predictor, "count")) + 
    geom_boxplot() + scale_y_continuous(trans='log10') + 
    theme_bw() + xlab("") + ylab("fp counts") + theme(axis.text.x = element_text(angle = 90, hjust = 1))
  # 3nt plots
  pred_barplot_3nt <- ggplot(resid_table_3nt, aes_string(predictor)) + geom_bar() + 
    theme_bw() + xlab("") + ylab("# obs") + ggtitle("") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1))
  pred_residplot_3nt <- ggplot(resid_table_3nt, aes_string(predictor, "residual")) + 
    geom_boxplot(aes(color=fit)) + theme_bw() + xlab("") + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1))
  pred_counts_3nt <- ggplot(resid_table_3nt, aes_string(predictor, "count")) + 
    geom_boxplot() + scale_y_continuous(trans='log10') + 
    theme_bw() + xlab("") + ylab("fp counts") + theme(axis.text.x = element_text(angle = 90, hjust = 1))
  # combine plots
  pred_barplot_2nt + pred_barplot_3nt + 
    pred_residplot_2nt + pred_residplot_3nt +
    pred_counts_2nt + pred_counts_3nt +
    plot_layout(heights=c(2,3,2), ncol=2)
}

plot_residuals_predictors(residuals_2nt, residuals_3nt, "f5", "5' bias sequence")
plot_residuals_predictors(residuals_2nt, residuals_3nt, "f3", "3' bias sequence")
plot_residuals_predictors(residuals_2nt, residuals_3nt, "A", "A site codon")
plot_residuals_predictors(residuals_2nt, residuals_3nt, "P", "P site codon")
plot_residuals_predictors(residuals_2nt, residuals_3nt, "E", "E site codon")
plot_residuals_predictors(residuals_2nt, residuals_3nt, "transcript", "transcript")
plot_residuals_predictors(residuals_2nt, residuals_3nt, "d5", "5' digest length")
plot_residuals_predictors(residuals_2nt, residuals_3nt, "d3", "3' digest length")
```

* 5' bias sequence: models predict poorly on AC, AT, and TC sequences

+ Otherwise, pretty good prediction (i.e. residuals mostly 0) on other sequences

* Magnitude of residuals doesn't tend to correlate with number of observations or footprint counts

### compare regression coefficients and p-values between models

```{r compare-models}
coef_3nt <- data.frame(sapply(all_fits_3nt, coef))
pval_3nt <- data.frame(sapply(all_fits_3nt, function(x) coef(summary(x))[,4]))
eval_3nt <- rbind(coef_3nt, pval_3nt)
eval_3nt$value <- c(rep("coef", nrow(coef_3nt)), rep("pval", nrow(pval_3nt)))
eval_3nt$type <- ""
eval_3nt$type[grepl("^(A|P|E)", rownames(eval_3nt))] <- "codon"
eval_3nt$type[grepl("^d", rownames(eval_3nt))] <- "digest length"
eval_3nt$type[grepl("^f", rownames(eval_3nt))] <- "sequence bias"
eval_3nt_quasipois <- eval_3nt[,c("poisson", "quasipois", "value", "type")]
eval_3nt_quasipois$fit <- "Quasi-poisson"
colnames(eval_3nt_quasipois)[2] <- "other"
eval_3nt_negbin <- eval_3nt[,c("poisson", "negbin", "value", "type")]
eval_3nt_negbin$fit <- "Negative binomial"
colnames(eval_3nt_negbin)[2] <- "other"
eval_3nt <- rbind(eval_3nt_quasipois, eval_3nt_negbin)
eval_3nt$fit <- factor(eval_3nt$fit, levels=c("Quasi-poisson", "Negative binomial"))
eval_3nt_coef <- ggplot(subset(eval_3nt, value=="coef"), aes(x=poisson, y=other)) + 
  geom_point(aes(color=type), alpha=0.5) + geom_abline(intercept=0, slope=1) +
  facet_grid(~fit, scales="free") + theme_bw() + xlab("Poisson") + ylab("") + ggtitle("Regression coefficients")
eval_3nt_pval <- ggplot(subset(eval_3nt, value=="pval"), aes(x=poisson, y=other)) + 
  geom_point(aes(color=type), alpha=0.5) + geom_abline(intercept=0, slope=1) +
  facet_grid(~fit, scales="free") + theme_bw() + xlab("Poisson") + ylab("") + ggtitle("p-values")
eval_3nt_coef / eval_3nt_pval
```

### compare A site regression coefficients to simulation parameters

```{r Asite-regression-vs-simulation}
# load A site simulation probabilities
codons <- apply(expand.grid(c("A", "T", "C", "G"),
                            c("A", "T", "C", "G"),
                            c("A", "T", "C", "G")),
                1, paste, collapse="")
weinberg_codonTEfile <- "~/simRiboSeq/refData/tunney_supp_table_2_codon_scores.csv"
weinberg_codonScores <- read.csv(file.path(weinberg_codonTEfile),
                                 header=T, stringsAsFactors=F)
weinberg_codonTE <- weinberg_codonScores$X0
weinberg_codonTE <- weinberg_codonTE - min(weinberg_codonTE) # scale up so min = 0.1
weinberg_stopCodons <- codons[which(!(codons %in% weinberg_codonScores$codon))]
weinberg_codonTE <- c(weinberg_codonTE, rep(0, length(weinberg_stopCodons))) # add TE for stop codons
names(weinberg_codonTE) <- c(weinberg_codonScores$codon, weinberg_stopCodons)
# scale probabilities relative to reference level
A_ref <- levels(regression_data_2nt$A)[1]
weinberg_codonTE <- weinberg_codonTE/weinberg_codonTE[A_ref]
# codon frequencies in transcripts
transcripts <- readLines("../sample_data/riboViz_transcripts_18cds15.fa")
transcript_names <- sub(">", "", grep(">", transcripts, value=T))
transcripts <- transcripts[!grepl(">", transcripts)]
names(transcripts) <- transcript_names
transcripts <- sapply(transcripts, function(x) substr(x, 18+1, nchar(x)-15))
transcripts <- strsplit(transcripts, split="")
transcripts_codons <- lapply(transcripts, 
                             function(x) {
                               paste0(x[c(T, F, F)],
                                      x[c(F, T, F)],
                                      x[c(F, F, T)])
                             })
transcripts_codons_cts <- sapply(transcripts_codons,
                                 function(x) {
                                   sapply(codons,
                                          function(y) {
                                            sum(y == x)
                                          })
                                 })
# create data frame for ggplot
codons_subset <- sort(unique(unlist(transcripts_codons)))
Asite_params<- data.frame(codon=c(levels(residuals_2nt$A)[-1],
                                  levels(residuals_3nt$A)[-1]),
                          sim=(weinberg_codonTE[match(codons_subset, names(weinberg_codonTE))])[-1],
                          beta=c(unlist(sapply(all_fits_2nt,
                                               function(x) coef(x)[grepl("^A", names(coef(x)))])),
                                 unlist(sapply(all_fits_3nt,
                                               function(x) coef(x)[grepl("^A", names(coef(x)))]))),
                          fit=c(rep(names(all_fits_2nt), times=length(codons_subset)-1),
                                rep(names(all_fits_3nt), times=length(codons_subset)-1)),
                          model=c(rep("2nt", times=length(all_fits_2nt)*(length(codons_subset)-1)),
                                  rep("3nt", times=length(all_fits_3nt)*(length(codons_subset)-1))))
Asite_params$logFC <- exp(Asite_params$beta)
## correlation between simulation and regression parameters
fit_model <- expand.grid(levels(Asite_params$fit), levels(Asite_params$model), stringsAsFactors=F)
colnames(fit_model) <- c("fit", "model")
fit_model$cor <- sapply(1:nrow(fit_model),
                        function(x) {
                          tmp_fit <- fit_model$fit[x]
                          tmp_model <- fit_model$model[x]
                          dat <- subset(Asite_params, fit==tmp_fit & model==tmp_model)
                          cor(dat$sim, dat$logFC)
                        })
Asite_params$fit <- sapply(1:nrow(Asite_params),
                           function(x) {
                             tmp_fit <- Asite_params$fit[x]
                             tmp_model <- Asite_params$model[x]
                             tmp_cor <- fit_model$cor[fit_model$fit==tmp_fit &
                                                        fit_model$model==tmp_model]
                             paste0(paste(tmp_fit, tmp_model), " ; cor=", signif(tmp_cor, 3))
                           })
# plot regression parameters
Asite <- ggplot(Asite_params, aes(x=sim, y=logFC)) + geom_point(aes(color=fit)) + 
  geom_abline(intercept=0, slope=1) + facet_grid(~fit) +
  theme_bw() + theme(legend.position="none") + 
  xlab("simulation logFC") + ylab("regression logFC")
# plot correlations
fit_model$fit <- factor(fit_model$fit, levels=c("poisson", "quasipois", "negbin"))
fit_model$model <- factor(fit_model$model, levels=c("2nt", "3nt"))
Asite_cor <- ggplot(fit_model) + geom_bar(aes(x=fit, y=cor, fill=fit), stat="identity") +
  facet_grid(~model) + theme_bw() + ylab("correlation") + ylim(0,1)
```

### plot all codon parameters

```{r APE_codons, fig.height=12}
codon_params <- rbind(do.call(rbind, lapply(1:length(all_fits_2nt),
                                            function(x) {
                                              fit <- all_fits_2nt[[x]]
                                              coefs <- coef(fit)[grepl("^(A|P|E)", names(coef(fit)))]
                                              data.frame(site=substr(names(coefs), start=1, stop=1),
                                                         codon=substr(names(coefs), start=2, stop=4),
                                                         param=coefs,
                                                         fit=rep(names(all_fits_2nt)[x], times=length(coefs)),
                                                         model="2nt")
                                            })),
                      do.call(rbind, lapply(1:length(all_fits_3nt),
                                            function(x) {
                                              fit <- all_fits_3nt[[x]]
                                              coefs <- coef(fit)[grepl("^(A|P|E)", names(coef(fit)))]
                                              data.frame(site=substr(names(coefs), start=1, stop=1),
                                                         codon=substr(names(coefs), start=2, stop=4),
                                                         param=coefs,
                                                         fit=rep(names(all_fits_3nt)[x], times=length(coefs)),
                                                         model="3nt")
                                            })))
codons_plot_all <- ggplot(codon_params, aes(x=site, y=param)) + geom_violin(aes(fill=fit)) + 
  facet_grid(model~fit) + theme_bw() + xlab("codon site") + ylab("regression coefficient") + 
  theme(legend.position = "none") + ggtitle("(full range)")
codons_plot_subset <- ggplot(codon_params, aes(x=site, y=param)) + geom_violin(aes(fill=fit)) + 
  facet_grid(model~fit) + ylim(-5,2) + ggtitle("(zoomed in)") +
  theme_bw() + xlab("codon site") + ylab("regression coefficient") + theme(legend.position = "none")
codons_plot_all / codons_plot_subset
```

* Data were simulated such that ribosome positions only depend on A site codon identity

+ Regression coefficients for E and P site codons should be 0

* Among 2nt models, negative binomial regression resulted in more E and P site regression coefficients at 0

* All 3nt models tend to have E and P site regression coefficients around 0

+ Regression coefficients for negative binomial model tighter around 0

### compare bias sequence regression coefficients to simulation parameters

```{r bias-regression-vs-simulation-f5}
# load bias sequence probabilities
green_biasFile <- "~/simRiboSeq/refData/codon_scores.tsv"
green_biasScores <- read.table(green_biasFile)
colnames(green_biasScores) <- as.character(seq.int(from=-5, length.out=ncol(green_biasScores)))
green_p5bias <- exp(green_biasScores[,"-5"])
names(green_p5bias) <- sort(codons)
green_p5bias <- green_p5bias/max(green_p5bias, na.rm=T)
green_p5bias[is.na(green_p5bias)] <- 0
green_n3bias <- exp(green_biasScores[,"3"])
names(green_n3bias) <- sort(codons)
green_n3bias <- green_n3bias/max(green_n3bias, na.rm=T)
green_n3bias[is.na(green_n3bias)] <- 0
# scale probabilities to reference level
p5_2nt_ref <- mean(green_p5bias[grepl(paste0("^", levels(regression_data_2nt$f5)[1]), names(green_p5bias))])
p5_3nt_ref <- green_p5bias[levels(regression_data_3nt$f5)[1]]
n3_2nt_ref <- mean(green_n3bias[grepl(paste0("^", levels(regression_data_3nt$f3)[1]), names(green_n3bias))])
n3_3nt_ref <- green_n3bias[levels(regression_data_3nt$f3)[1]]
green_p5bias_2nt <- green_p5bias/p5_2nt_ref
green_p5bias_3nt <- green_p5bias/p5_3nt_ref
green_n3bias_2nt <- green_n3bias/n3_2nt_ref
green_n3bias_3nt <- green_n3bias/n3_3nt_ref
# create data frame from ggplot (2nt)
bias_params_2nt <- do.call(rbind, lapply(1:length(all_fits_2nt),
                                         function(x) {
                                           coefs <- coef(all_fits_2nt[[x]])
                                           coefs <- coefs[grepl("^f", names(coefs))]
                                           data.frame(region=substr(names(coefs), start=1, stop=2),
                                                      seq=substr(names(coefs), start=3, stop=4),
                                                      param=coefs,
                                                      fit=names(all_fits_2nt),
                                                      model="2nt")
                                         }))
bias_params_2nt <- reshape::expand.grid.df(bias_params_2nt, data.frame(end=c("A", "T", "C", "G")))
bias_params_2nt$sim <- sapply(1:nrow(bias_params_2nt),
                              function(x) {
                                tmp_seq <- paste0(bias_params_2nt$seq[x], bias_params_2nt$end[x])
                                ifelse(bias_params_2nt$region[x]=="f5",
                                       green_p5bias_2nt[match(tmp_seq, names(green_p5bias_2nt))],
                                       green_n3bias_2nt[match(tmp_seq, names(green_n3bias_2nt))])
                              })
bias_params_2nt <- subset(bias_params_2nt, select=-c(end))
# create data frame from ggplot (3nt)
bias_params_3nt <- do.call(rbind, lapply(1:length(all_fits_3nt),
                                         function(x) {
                                           coefs <- coef(all_fits_3nt[[x]])
                                           coefs <- coefs[grepl("^f", names(coefs))]
                                           data.frame(region=substr(names(coefs), start=1, stop=2),
                                                      seq=substr(names(coefs), start=3, stop=5),
                                                      param=coefs,
                                                      fit=names(all_fits_3nt)[x],
                                                      model="3nt")
                                         }))
bias_params_3nt$sim <- sapply(1:nrow(bias_params_3nt),
                              function(x) {
                                ifelse(bias_params_3nt$region[x]=="f5",
                                       green_p5bias_3nt[match(bias_params_3nt$seq[x], names(green_p5bias_3nt))],
                                       green_n3bias_3nt[match(bias_params_3nt$seq[x], names(green_n3bias_3nt))])
                              })
bias_params_3nt$nObs <- sapply(1:nrow(bias_params_3nt),
                               function(x) {
                                 tmp_seq <- as.character(bias_params_3nt$seq[x])
                                 ifelse(bias_params_3nt$region[x]=="f5",
                                        sum(as.character(regression_data_3nt$f5)==tmp_seq),
                                        sum(as.character(regression_data_3nt$f3)==tmp_seq))
                               })
bias_params_3nt$subseq <- sapply(1:nrow(bias_params_3nt),
                                 function(x) {
                                   ifelse(bias_params_3nt$region[x]=="f5",
                                          substr(bias_params_3nt$seq[x], start=1, stop=2),
                                          substr(bias_params_3nt$seq[x], start=2, stop=3))
                                 })
# combine data frames and plot
bias_params <- rbind(bias_params_2nt, bias_params_3nt)
bias_params$region <- factor(bias_params$region, levels=c("f5", "f3"))
bias_params$logFC <- exp(bias_params$param)
ggplot(subset(bias_params, region=="f5"), aes(x=sim, y=logFC)) + geom_point(aes(color=seq)) + 
  facet_grid(~ model + fit) + theme_bw() + 
  xlab("simulation parameter") + ylab("regression coefficient") + ggtitle("5' sequence bias")
for(i in 1:length(all_fits_3nt)) {
  f5_coefs <- coef(all_fits_3nt[[i]])[grepl("^f5", names(coef(all_fits_3nt[[i]])))]
  names(f5_coefs) <- sub("f5", "", names(f5_coefs))
  f5_coefs_outliers <- f5_coefs[f5_coefs<(-10)]
  print(paste(names(all_fits_3nt)[i], ":", paste(names(f5_coefs_outliers), collapse=",")))
}
```

* 2nt: plotted 4 simulation parameters per bias sequence (i.e. for AA: plotted AAA, AAT, AAC, AAG)

+ Mostly linear correlation between regression coefficient and simulation probability

* 3nt: consistently 3 outliers (TAA, TAG, TGA)

+ Otherwise, good linear fit between regression coefficient and simulation probability

+ Outliers worse in negative binomial fit

+ Number of observations: TAA (441), TAG (171), TGA (1476)

```{r bias-regression-vs-simulation-f3}
ggplot(subset(bias_params, region=="f3"), aes(x=sim, y=param)) + geom_point(aes(color=seq)) + 
  facet_grid(~ model + fit) + theme_bw() + 
  xlab("simulation parameter") + ylab("regression coefficient") + ggtitle("3' sequence bias")
```

* Regression coefficients tend to be 0

* Need to go back into simulation code to check for errors

### bias correction

* move forward with 3 nt negative binomial model

1. set all 5' and 3' bias sequences to the reference level (AAA)

2. predict footprint counts with modified data

3. codon correlation plots to evaluate bias correction

+ collapse to counts per codon

+ scale footprint counts: divide raw counts by average across transcript

+ full model: -7 to +5 codons; plus 13 leave-one-out models

+ compute Pearson correlations btwn true & predicted scaled counts