---
title: "Scalability Evaluation: simASite"
author: "Amanda Mok"
date: "1/29/2020"
output: html_document
---

```{r set-up, echo=F}
knitr::opts_chunk$set(cache=T, echo=F, fig.width=12, warning=F)

library(ggplot2)
library(patchwork)

source("~/footprint-bias/scripts/correct_bias.R")

scalability_dir <- "~/footprint-bias/scalability_simASite/"
for(x in grep("Rda", list.files(scalability_dir, recursive=T), value=T)) {load(file.path(scalability_dir, x))}

n <- sort(as.numeric(sub("genes_", "", list.files(scalability_dir))))
n <- n[-length(n)]
```

# computing time

```{r timing}
timing <- data.frame(nGenes = n,
                     nRows = sapply(n, function(x) nrow(get(paste0("regression_data_", x)))),
                     time = c(sapply(n, function(x) get(paste("time", x, "init", sep="_"))$min),
                              sapply(n, function(x) get(paste("time", x, "count", sep="_"))$min),
                              sapply(n, function(x) get(paste("time", x, "regression", sep="_"))$min)),
                     type = rep(c("initialize", "count", "regression"), each=length(n)))
timing$type <- factor(timing$type, levels=c("initialize", "count", "regression"))
timing$time <- timing$time / 60
ggplot(timing, aes(x=nRows, y=time)) + geom_line() + geom_text(aes(label=nGenes)) + facet_grid(~type) +
  theme_bw() + ylab("time (min)") + ggtitle("compute time") + 
  theme(axis.text.x=element_text(angle=90))
```

# memory allocation

```{r memory}
memory <- data.frame(nGenes = n,
                     nRows = sapply(n, function(x) nrow(get(paste0("regression_data_", x)))),
                     mem = c(sapply(n, function(x) get(paste("time", x, "init", sep="_"))$mem_alloc),
                             sapply(n, function(x) get(paste("time", x, "count", sep="_"))$mem_alloc),
                             sapply(n, function(x) get(paste("time", x, "regression", sep="_"))$mem_alloc)),
                     type = rep(c("initialize", "count", "regression"), each=length(n)))
memory$type <- factor(timing$type, levels=c("initialize", "count", "regression"))
ggplot(memory, aes(x=nRows, y=mem)) + geom_line() + geom_text(aes(label=nGenes)) + facet_grid(~type) +
  theme_bw() + ylab("memory allocated (bytes)") + ggtitle("memory allocation") +
  theme(axis.text.x=element_text(angle=90)) + scale_y_log10()
```

# regression coefficients

```{r coef, fig.height=12}
# make data.frame for ggplot()
all_coefs <- lapply(n,
                    function(x) {
                      tmp_fit <- get(paste0("nb_fit_", x))
                      if("try-error" %in% class(tmp_fit)) {
                        print(paste(x, "try-error"))
                        return(NULL)
                      } else {
                        tmp_coef <- coef(tmp_fit)
                        nRows <- nrow(get(paste0("regression_data_", x)))
                        return(data.frame(value=tmp_coef,
                                          coef=names(tmp_coef),
                                          nGenes=x,
                                          nRows=nRows))
                      }
                    })
all_coefs <- do.call(rbind, all_coefs)
all_coefs$type <- ""
all_coefs$type[grepl("^transcript", all_coefs$coef)] <- "transcript"
all_coefs$type[grepl("^(A|P|E)", all_coefs$coef)] <- "codon"
all_coefs$type[grepl("^d", all_coefs$coef)] <- "digest"
all_coefs$type[grepl("^f", all_coefs$coef)] <- "bias"
all_coefs$subtype <- ""
subtypes <- c("A", "P", "E", "d5", "d3", "f5", "f3")
for(x in subtypes) {
  all_coefs$subtype[grepl(paste0("^", x), all_coefs$coef)] <- x
}
all_coefs$subtype <- factor(all_coefs$subtype, levels=subtypes)
all_coefs$exp <- exp(all_coefs$value)

# make plots
coef_transcript <- ggplot(subset(all_coefs, type=="transcript"), aes(x=nRows, y=exp, color=coef)) +
  geom_line() + geom_point() + theme_bw() + xlab("") + ylab(expression("exp("*beta*")")) +
  theme(legend.position="none", axis.text.x=element_text(angle=90)) + ggtitle("transcript")
coef_codon <- ggplot(subset(all_coefs, type=="codon"), aes(x=nRows, y=exp, color=coef)) + 
  geom_line() + geom_point() + theme_bw() + xlab("") + ylab(expression("exp("*beta*")")) + 
  theme(legend.position="none", axis.text.x=element_text(angle=90)) + ggtitle("codons") + facet_grid(~subtype)
coef_digest <- ggplot(subset(all_coefs, type=="digest"), aes(x=nRows, y=exp, color=coef)) + 
  geom_line() + geom_point() + theme_bw() + xlab("") + ylab(expression("exp("*beta*")")) +
  theme(legend.position="none", axis.text.x=element_text(angle=90)) + ggtitle("digest lengths")
coef_bias <- ggplot(subset(all_coefs, type=="bias"), aes(x=nRows, y=exp, color=coef)) +
  geom_line() + geom_point() + theme_bw() + xlab("number of rows") + ylab(expression("exp("*beta*")")) +
  theme(legend.position="none", axis.text.x=element_text(angle=90)) + ggtitle("sequence bias") +
  facet_grid(~subtype)
(coef_transcript + coef_digest) / coef_codon / coef_bias
```

# evaluate bias correction

```{r eval_bias, fig.height=12}
all_bias_plots <- lapply(n,
                         function(x) {
                           is_error <- class(get(paste0("nb_fit_", x)))
                           if("try-error" %in% is_error) {
                             return(NULL)
                           } else {
                             tmp_corrected <- correct_bias(get(paste0("regression_data_", x)), 
                                                           get(paste0("nb_fit_", x)))
                             tmp_corrected <- count_by_codon(tmp_corrected)
                             tmp_transcripts_fa_fname <- file.path(scalability_dir, 
                                                                   paste0("genes_", x), 
                                                                   paste0("scer_", x, "_transcripts_18cds15.fa"))
                             tmp_bias <- evaluate_bias(tmp_corrected, tmp_transcripts_fa_fname)
                             tmp_plot <- plot_bias(tmp_bias, plotTitle=paste(x, "genes"))
                             return(tmp_plot)
                           }
                         })
do.call(wrap_plots, list(all_bias_plots[!sapply(all_bias_plots, is.null)], ncol=1))
```